lib/util/Components.js

```diff
@@ -10,164 +10,167 @@ const doctrine = require('doctrine');
+      let node = null;
+      node = this._list[i].node;
+      while (!component && node.parent) {
+        node = node.parent;
+        // Stop moving up if we reach a decorator
+        if (node.type === 'Decorator') {
+          break;
+        }
+        component = this.get(node);
+      }
+      if (component) {
+        const newUsedProps = (this._list[i].usedPropTypes || []).filter(propType => !propType.node || propType.node.kind !== 'init');
+
+        const componentId = this._getId(component.node);
+        usedPropTypes[componentId] = (usedPropTypes[componentId] || []).concat(newUsedProps);
       }
-      component = this.get(node);
     }
-    if (component) {
-      const newUsedProps = (this._list[i].usedPropTypes || []).filter(propType => !propType.node || propType.node.kind !== 'init');
 
-      const componentId = this._getId(component.node);
-      usedPropTypes[componentId] = (usedPropTypes[componentId] || []).concat(newUsedProps);
+    // Assign used props in not confident components to the parent component
+    for (const j in this._list) {
+      if (!has(this._list, j) || this._list[j].confidence < 2) {
+        continue;
+      }
+      const id = this._getId(this._list[j].node);
+      list[j] = this._list[j];
+      if (usedPropTypes[id]) {
+        list[j].usedPropTypes = (list[j].usedPropTypes || []).concat(usedPropTypes[id]);
+      }
     }
+    return list;
   }
-  // Assign used props in not confident components to the parent component
-  for (const j in this._list) {
-    if (!has(this._list, j) || this._list[j].confidence < 2) {
-      continue;
-    }
-    const id = this._getId(this._list[j].node);
-    list[j] = this._list[j];
-    if (usedPropTypes[id]) {
-      list[j].usedPropTypes = (list[j].usedPropTypes || []).concat(usedPropTypes[id]);
+
+  /**
+   * Return the length of the components list
+   * Components for which we are not confident are not counted
+   *
+   * @returns {Number} Components list length
+   */
+  length() {
+    let length = 0;
+    for (const i in this._list) {
+      if (!has(this._list, i) || this._list[i].confidence < 2) {
+        continue;
+      }
+      length++;
     }
+    return length;
   }
-  return list;
-};
 
-/**
- * Return the length of the components list
- * Components for which we are not confident are not counted
- *
- * @returns {Number} Components list length
- */
-Components.prototype.length = function() {
-  let length = 0;
-  for (const i in this._list) {
-    if (!has(this._list, i) || this._list[i].confidence < 2) {
-      continue;
+  _mergeUsedPropTypes(propsList, newPropsList) {
+    const propsToAdd = [];
+    newPropsList.forEach(newProp => {
+      const newPropisAlreadyInTheList = propsList.some(prop => this._usedPropTypesAreEquivalent(prop, newProp));
+      if (!newPropisAlreadyInTheList) {
+        propsToAdd.push(newProp);
+      }
+    });
+    return propsList.concat(propsToAdd);
+  }
+
+  _usedPropTypesAreEquivalent(propA, propB) {
```

Comment:
"_" doesn't make anything private; this change makes them fully public. Let's revert these back to functions so that the available prototype methods don't change.

Reply:
Okay, should probably do the same for `_getId` then?

Reply:
yes please